File: mockserver-client-dotnet/src/MockServer.Client/Models/DnsResponse.cs
using System.Text.Json.Serialization;

namespace MockServer.Client.Models;

/// <summary>
/// A single DNS resource record within a <see cref="DnsResponse"/>.
/// </summary>
public sealed class DnsRecord
{
    [JsonPropertyName("name")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Name { get; set; }

    [JsonPropertyName("type")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Type { get; set; }

    [JsonPropertyName("dnsClass")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? DnsClass { get; set; }

    [JsonPropertyName("ttl")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Ttl { get; set; }

    [JsonPropertyName("value")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Value { get; set; }

    [JsonPropertyName("priority")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Priority { get; set; }

    [JsonPropertyName("weight")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Weight { get; set; }

    [JsonPropertyName("port")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Port { get; set; }

    /// <summary>An IPv4 address (A) record.</summary>
    public static DnsRecord ARecord(string name, string ip) => new() { Name = name, Type = "A", Value = ip };

    /// <summary>An IPv6 address (AAAA) record.</summary>
    public static DnsRecord AaaaRecord(string name, string ip) => new() { Name = name, Type = "AAAA", Value = ip };

    /// <summary>A canonical name (CNAME) record.</summary>
    public static DnsRecord CnameRecord(string name, string cname) => new() { Name = name, Type = "CNAME", Value = cname };

    /// <summary>A mail exchanger (MX) record.</summary>
    public static DnsRecord MxRecord(string name, int priority, string exchange)
        => new() { Name = name, Type = "MX", Priority = priority, Value = exchange };

    /// <summary>A service locator (SRV) record.</summary>
    public static DnsRecord SrvRecord(string name, int priority, int weight, int port, string target)
        => new() { Name = name, Type = "SRV", Priority = priority, Weight = weight, Port = port, Value = target };

    /// <summary>A text (TXT) record.</summary>
    public static DnsRecord TxtRecord(string name, string text) => new() { Name = name, Type = "TXT", Value = text };

    /// <summary>A pointer (PTR) record.</summary>
    public static DnsRecord PtrRecord(string name, string pointer) => new() { Name = name, Type = "PTR", Value = pointer };
}

/// <summary>
/// Represents a DNS response action for MockServer. When MockServer is acting
/// as a DNS server, the configured records are returned for matched queries.
/// </summary>
public sealed class DnsResponse
{
    [JsonPropertyName("responseCode")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? ResponseCode { get; set; }

    [JsonPropertyName("answerRecords")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<DnsRecord>? AnswerRecords { get; set; }

    [JsonPropertyName("authorityRecords")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<DnsRecord>? AuthorityRecords { get; set; }

    [JsonPropertyName("additionalRecords")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<DnsRecord>? AdditionalRecords { get; set; }

    [JsonPropertyName("delay")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public Delay? Delay { get; set; }

    [JsonPropertyName("primary")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? Primary { get; set; }

    /// <summary>
    /// Creates a new DNS response builder.
    /// </summary>
    public static DnsResponseBuilder Response() => new();
}

File: mockserver-client-php/src/DnsResponse.php
<?php

declare(strict_types=1);

namespace MockServer;

/**
 * Fluent builder for a DNS response action.
 *
 * Produces the {@code dnsResponse} action JSON. Wire keys:
 * responseCode, answerRecords, authorityRecords, additionalRecords, delay, primary.
 *
 * @example
 *   DnsResponse::response()
 *       ->responseCode('NOERROR')
 *       ->answer(DnsRecord::aRecord('example.com', '1.2.3.4'));
 */
class DnsResponse implements \JsonSerializable
{
    private ?string $responseCode = null;
    /** @var list<DnsRecord> */
    private array $answerRecords = [];
    /** @var list<DnsRecord> */
    private array $authorityRecords = [];
    /** @var list<DnsRecord> */
    private array $additionalRecords = [];
    private ?Delay $delay = null;
    private ?bool $primary = null;

    public static function response(): self
    {
        return new self();
    }

    public function responseCode(string $responseCode): self
    {
        $this->responseCode = $responseCode;
        return $this;
    }

    /**
     * Append an answer-section record.
     */
    public function answer(DnsRecord $record): self
    {
        $this->answerRecords[] = $record;
        return $this;
    }

    /**
     * Append an authority-section record.
     */
    public function authority(DnsRecord $record): self
    {
        $this->authorityRecords[] = $record;
        return $this;
    }

    /**
     * Append an additional-section record.
     */
    public function additional(DnsRecord $record): self
    {
        $this->additionalRecords[] = $record;
        return $this;
    }

    public function delay(Delay $delay): self
    {
        $this->delay = $delay;
        return $this;
    }

    public function primary(bool $primary): self
    {
        $this->primary = $primary;
        return $this;
    }

    /**
     * @return array<string, mixed>
     */
    public function jsonSerialize(): array
    {
        return $this->toArray();
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $data = [];
        if ($this->responseCode !== null) {
            $data['responseCode'] = $this->responseCode;
        }
        if (!empty($this->answerRecords)) {
            $data['answerRecords'] = array_map(fn(DnsRecord $r) => $r->toArray(), $this->answerRecords);
        }
        if (!empty($this->authorityRecords)) {
            $data['authorityRecords'] = array_map(fn(DnsRecord $r) => $r->toArray(), $this->authorityRecords);
        }
        if (!empty($this->additionalRecords)) {
            $data['additionalRecords'] = array_map(fn(DnsRecord $r) => $r->toArray(), $this->additionalRecords);
        }
        if ($this->delay !== null) {
            $data['delay'] = $this->delay->toArray();
        }
        if ($this->primary !== null) {
            $data['primary'] = $this->primary;
        }
        return $data;
    }
}
